lib/PrivateAppUserTokenManager.ts

```diff
@@ -0,0 +1,243 @@
+import {
+  fetchPrivateAppUserToken,
+  createPrivateAppUserToken,
+  updatePrivateAppUserToken,
+  PrivateAppUserTokenResponse,
+} from '../api/privateAppUserToken';
+import { scopeGroupsForPersonalAccessKey } from './personalAccessKey';
+import { logger } from './logger';
+import { i18n } from '../utils/lang';
+import moment from 'moment';
+import { AxiosError, isAxiosError } from 'axios';
+import { getAxiosErrorWithContext } from '../errors/apiErrors';
+import { throwErrorWithMessage } from '../errors/standardErrors';
+import { BaseError } from '../types/Error';
+
+const USER_TOKEN_READ = 'developer.private_app.temporary_token.read';
+const USER_TOKEN_WRITE = 'developer.private_app.temporary_token.write';
+const i18nKey = 'lib.PrivateAppUserTokenManager';
+
+export class PrivateAppUserTokenManager {
+  accountId: number;
+  private tokenMap: Map<number, PrivateAppUserTokenResponse>;
+  private tokenMapIntervalId: Map<number, NodeJS.Timeout>;
+  private enabled: boolean;
+
+  constructor(accountId: number) {
+    this.accountId = accountId;
+    this.tokenMap = new Map<number, PrivateAppUserTokenResponse>();
+    this.tokenMapIntervalId = new Map<number, NodeJS.Timeout>();
+    this.enabled = false;
+  }
+
+  async init(): Promise<void> {
+    const scopeGroups = new Set<string>(
+      await scopeGroupsForPersonalAccessKey(this.accountId)
+    );
+    if (
+      !scopeGroups.has(USER_TOKEN_READ) ||
+      !scopeGroups.has(USER_TOKEN_WRITE)
+    ) {
+      throwErrorWithMessage(`${i18nKey}.errors.noScopes`, {
+        accountId: this.accountId,
+      });
+    } else {
+      logger.debug(
+        i18n(`${i18nKey}.enabled`, {
+          accountId: this.accountId,
+        })
+      );
+      this.enabled = true;
+    }
+  }
+
+  cleanup() {
+    this.tokenMapIntervalId.forEach(timeoutId => clearInterval(timeoutId));
+    this.tokenMapIntervalId.clear();
+    this.tokenMap.clear();
+  }
+
+  async getPrivateAppToken(
+    appId: number,
+    scopeGroups: string[] = []
+  ): Promise<string | undefined> {
+    if (!this.enabled) {
+      logger.debug(
+        i18n(`${i18nKey}.disabled`, {
+          accountId: this.accountId,
+          appId: appId,
+        })
+      );
+      return;
+    }
+    try {
+      if (
+        this.tokenMap.has(appId) &&
+        PrivateAppUserTokenManager.doesTokenHaveAllScopes(
+          this.tokenMap.get(appId),
+          scopeGroups
+        )
+      ) {
+        logger.debug(
+          i18n(`${i18nKey}.cached`, {
+            appId: appId,
+          })
+        );
+        return this.tokenMap.get(appId)!.userTokenKey;
+      } else {
+        const token = await this.createOrGetActiveToken(appId, scopeGroups);
+        this.setCacheAndRefresh(appId, token, scopeGroups);
```

Comment:
`token` includes `token.scopeGroups` right? I'm thinking that we might be able to cut down on args in some places by accessing `token.scopeGroups` rather than passing around this `scopeGroups` variable. `setCacheAndRefresh()` for example, is only ever called when it's passed a new token.

Reply:
left the variable for now; if we can grab the scope groups from the app.json file and pass it into the fetch it'll make sure that the token always has those scopes. The token's scope and the app json might be mismatched, though I can just delete all the SG stuff for now if we don't think we'll need it some point later